toucan_connectors/hubspot_private_app/hubspot_connector.py

```diff
@@ -0,0 +1,148 @@
+    paging: _HubSpotPaging | None = None
+    results: list[_HubSpotResult]
+
+    def next_page_after(self) -> str | None:
+        return None if self.paging is None else self.paging.next_.after
+
+
+_RawHubSpotResult: TypeAlias = dict[str, str | None | list[Any]]
+
+
+class _HubSpotObject(Protocol):
+    def to_dict(self) -> _RawHubSpotResult:
+        ...
+
+
+class _BasicApi(Protocol):
+    def get_page(self, after: str | None, limit: int | None) -> _HubSpotObject:
+        ...
+
+
+class _Api(Protocol):
+    @property
+    def basic_api(self) -> _BasicApi:
+        ...
+
+    def get_all(self) -> list[_HubSpotObject]:
+        ...
+
+
+class HubspotConnector(ToucanConnector):
+    data_source_model: HubspotDataSource
+    access_token: SecretStr = Field(..., description='An API key for the target private app')
+
+    def _fetch_page(
+        self, api: _Api, after: str | None = None, limit: int | None = None
+    ) -> _HubSpotResponse:
+        page = api.basic_api.get_page(after=after, limit=limit)
+        return _HubSpotResponse(**page.to_dict())
+
+    def _fetch_all(self, api: _Api) -> list[_HubSpotResult]:
+        results = api.get_all()
+        return [_HubSpotResult(**elem.to_dict()) for elem in results]
+
+    def _api_for_dataset(self, object_type: HubspotDataset) -> _Api:
+        client = HubSpot(access_token=self.access_token.get_secret_value())
+        return getattr(client.crm, object_type.value)
+
+    def _retrieve_data(self, data_source: HubspotDataSource) -> pd.DataFrame:
+        api = self._api_for_dataset(data_source.dataset)
+        return pd.DataFrame(r.to_dict() for r in self._fetch_all(api))
+
+    def _result_iterator(
+        self, api: _Api, max_results: int | None, limit: int | None
+    ) -> Generator[_HubSpotResult, None, None]:
+        after = None
+        count = 0
+        while True:
+            page = self._fetch_page(api, after=after, limit=limit)
+            for result in page.results:
+                if max_results is not None and count >= max_results:
+                    return
+                yield result
+                count += 1
+
+            if (after := page.next_page_after()) is None:
+                return
+
+    def get_slice(
+        self,
+        data_source: HubspotDataSource,
+        permissions: dict | None = None,
+        offset: int = 0,
+        limit: int | None = None,
+        get_row_count: bool | None = False,
+    ) -> DataSlice:
+        api = self._api_for_dataset(data_source.dataset)
+
+        if limit is None:
+            df = pd.DataFrame(self._fetch_all(api)[offset:])
+
+        else:
+            results = []
+            result_iterator = self._result_iterator(
+                api, max_results=offset + (limit or 0), limit=limit
+            )
+            try:
+                for _ in range(offset):
+                    next(result_iterator)
+                results = list(result_iterator)
+            except StopIteration:
+                pass
```

Comment:
Is it not better like this ? 🤔 
```suggestion
           except StopIteration:
                pass
           results = list(result_iterator)
```

If there is an exception, the results will not be set right ?

Reply:
If there's an exception, there's nothing left to read in the iterator, and `results` has already been assigned to an emtpy list on l.131